### ISSUE TITLE
IORING_OP_READ causes io_uring_enter to block if the fd is a terminal.

### ISSUE THREAD

**Issue by lonjil** [axboe/liburing#354]
I'm on kernel 5.12.8, but I observed the same behavior on 5.11.17.

```c
#include <stdio.h>
#include <liburing.h>

int main(void)
{
    struct io_uring ring;
    struct io_uring_sqe *sqe;
    struct io_uring_cqe *cqe;
    io_uring_queue_init(1, &ring, 0);
    sqe = io_uring_get_sqe(&ring);

    char buf[64] = {0};
    io_uring_prep_read(sqe, 0, buf, 64, -1);

    fprintf(stderr, "Submitting.\n");
    io_uring_submit(&ring);

    fprintf(stderr, "Waiting.\n");
    io_uring_wait_cqe(&ring, &cqe);
    fprintf(stderr, "{ user_data = %lld, res = %u, flags = %u\n", cqe->user_data, cqe->res, cqe->flags);
    io_uring_cqe_seen(&ring, cqe);
    
    return 0;
}
```
Running the above with `strace -e trace=io_uring_enter ./a.out` gives the following
```
Submitting.
io_uring_enter(3, 1, 0, 0, NULL, 8
```
Letting the read complete by hitting enter then results in
```
)     = 1
Waiting.
{ user_data = 0, res = 1, flags = 0
+++ exited with 0 +++
```

If instead, we do `cat | strace -e trace=io_uring_enter ./a.out`, we get
```
Submitting.
io_uring_enter(3, 1, 0, 0, NULL, 8)     = 1
Waiting.
io_uring_enter(3, 0, 1, IORING_ENTER_GETEVENTS, NULL, 8
```
and after hitting enter
```
) = 0
{ user_data = 0, res = 1, flags = 0
+++ exited with 0 +++
```

**Comment by isilence** [comment]
Not great. It's a char device and a kernel function `__io_file_supports_async()` let it slip through.

```C
if (S_ISCHR(mode) || S_ISSOCK(mode))
	return true;
```

Don't know why char devices are treated as this, need to take a look

**Comment by isilence** [comment]
@lonjil, to give attribution we tag commits with `Suggested-by: Full-Name <email@example.com>`
e.g. like here https://git.kernel.dk/cgit/linux-block/commit/?h=634d00df5e1cfc4a707b629a814bd607f726bd52

but I'd need your full name and email if you want.

**Comment by lonjil** [comment]
@isilence `Birk Hirdman <lonjil@gmail.com>`

**Comment by isilence** [comment]
Released with 5.14, will be backported to older kernels some time soon.
I think we can close it
